### PR TITLE
增加redisRegistry consumer侧的服务探活逻辑

### PR DESCRIPTION
**Pull request by lkxiaolou**
##  增加redisRegistry consumer侧的服务探活逻辑

目前redisRegistry的实现中，provider异常（未执行注销逻辑，如被kill -9）下线，只在`monitoring center`存在的情况下会对过期的provider进行下线操作，这有两个缺点：
- monitoring center不一定会部署
- 它的实现使用了keys，会阻塞redis； -> 更新：最新版本使用scan扫描，不会阻塞redis

## 解法：
在consumer侧对订阅到的服务进行检查，每隔 `1/2 session` 时间检查provider是否过期，如果过期，对它执行`doNotify`，即重新从redis获取最新数据，验证过期时间。

## 验证：
- 旧逻辑：使用redisRegistry，启动2个provider，1个consumer，对其中1个provider执行kill -9，consumer调用会有部分请求报错。
image

- 新逻辑：使用redisRegistry，启动2个provider，1个consumer，对其中1个provider执行kill -9，consumer会在`1/2 session`时间（默认30s，可配置）内将有问题的provider踢掉。
image

notify日志可以看出只有1个provider了

```
2021-05-31 10:52:45,312 [DubboRegistryExpireTimer-thread-1] INFO  org.apache.dubbo.registry.redis.RedisRegistry (RedisRegistry.java:472) -  [DUBBO] redis notify: /dubbo/com.didichuxing.basic.api.MyDemoService/providers = [dubbo://127.0.0.1:20880/com.didichuxing.basic.api.MyDemoService?anyhost=true&application=ddog-my-demo-p0&deprecated=false&dubbo=2.0.2&dynamic=true&generic=false&interface=com.didichuxing.basic.api.MyDemoService&methods=call&owner=roshilikang&pid=1543&release=2.7.6&side=provider&threads=500&timestamp=1622429433943], dubbo version: 2.7.8-SNAPSHOT, current host: 172.23.232.133
```

### CODE REVIEW

**Comment by AlbumenJ** [comment]
Would you please add some ut cases to verify this PR?

**Comment by lkxiaolou** [comment]
> Would you please add some ut cases to verify this PR?

done

**Comment by lkxiaolou** [comment]
@AlbumenJ @xukefu 1 workflow awaiting approval

**Comment by xukefu** [comment]
I don't know why you submitted `CompatibleTypeUtils`,this class you are currently submitting is the old one, and the new one has been merged into the master. If you merge your current PR directly, it will overwrite the optimized code and cause compilation errors.

**Comment by lkxiaolou** [comment]
> I don't know why you submitted `CompatibleTypeUtils`,this class you are currently submitting is the old one, and the new one has been merged into the master. If you merge your current PR directly, it will overwrite the optimized code and cause compilation errors.

File changed only `RedisRegistry` and `RedisRegistryTest`

File: dubbo-registry/dubbo-registry-redis/src/test/java/org/apache/dubbo/registry/redis/RedisRegistryTest.java [hunk 1]
```diff
@@ -105,6 +110,66 @@ public void testAnyHost() {
         });
     }
 
+    @Test
+    public void testSubscribeWhenProviderCrash() throws Exception {
+
+        // 关闭doExpire，则本单元测试不通过
```
Comment (AlbumenJ) [inline dubbo-registry/dubbo-registry-redis/src/test/java/org/apache/dubbo/registry/redis/RedisRegistryTest.java:116]: please comment in English

**Review by AlbumenJ — CHANGES_REQUESTED** [review]

File: dubbo-registry/dubbo-registry-redis/src/test/java/org/apache/dubbo/registry/redis/RedisRegistryTest.java (hunk 1 above)
Comment (lkxiaolou) [reply dubbo-registry/dubbo-registry-redis/src/test/java/org/apache/dubbo/registry/redis/RedisRegistryTest.java]: done

File: dubbo-registry/dubbo-registry-redis/src/main/java/org/apache/dubbo/registry/redis/RedisRegistry.java [hunk 2]
```diff
@@ -294,10 +308,12 @@ private void doNotify(Collection<String> keys, URL url, Collection<NotifyListene
             if (CollectionUtils.isNotEmptyMap(values)) {
                 for (Map.Entry<String, String> entry : values.entrySet()) {
                     URL u = URL.valueOf(entry.getKey());
+                    long expire = Long.parseLong(entry.getValue());
                     if (!u.getParameter(DYNAMIC_KEY, true)
-                            || Long.parseLong(entry.getValue()) >= now) {
+                            || expire >= now) {
                         if (UrlUtils.isMatch(url, u)) {
                             urls.add(u);
+                            expireCache.put(u, expire);
```
Comment (chickenlj) [inline dubbo-registry/dubbo-registry-redis/src/main/java/org/apache/dubbo/registry/redis/RedisRegistry.java:316]: Is it possible for expireCache to go leaking for it's never cleared?

**Review by chickenlj — CHANGES_REQUESTED** [review]

File: dubbo-registry/dubbo-registry-redis/src/main/java/org/apache/dubbo/registry/redis/RedisRegistry.java (hunk 2 above)
Comment (lkxiaolou) [reply dubbo-registry/dubbo-registry-redis/src/main/java/org/apache/dubbo/registry/redis/RedisRegistry.java:317]: ok , add remove logic and related UT

Comment (chickenlj) [reply dubbo-registry/dubbo-registry-redis/src/main/java/org/apache/dubbo/registry/redis/RedisRegistry.java:317]: Also, I don't see where `expireCache` is used inside `doNotify`.

Comment (lkxiaolou) [reply dubbo-registry/dubbo-registry-redis/src/main/java/org/apache/dubbo/registry/redis/RedisRegistry.java:317]: `expireCache` mark which service may be down and call `doNotity` to fetch latest data from redis

**Pull request merged**